src/storage/src/hummock/cache/block_cache.rs

```diff
@@ -0,0 +1,326 @@
+};
+
+const MIN_BUFFER_SIZE_PER_SHARD: usize = 32 * 1024 * 1024;
+
+enum BlockEntry {
+    Cache(CachableEntry<(HummockSstableId, u64), Box<Block>>),
+    Owned(Box<Block>),
+    RefEntry(Arc<Block>),
+}
+
+pub struct BlockHolder {
+    _handle: BlockEntry,
+    block: *const Block,
+}
+
+impl BlockHolder {
+    pub fn from_ref_block(block: Arc<Block>) -> Self {
+        let ptr = block.as_ref() as *const _;
+        Self {
+            _handle: BlockEntry::RefEntry(block),
+            block: ptr,
+        }
+    }
+
+    pub fn from_owned_block(block: Box<Block>) -> Self {
+        let ptr = block.as_ref() as *const _;
+        Self {
+            _handle: BlockEntry::Owned(block),
+            block: ptr,
+        }
+    }
+
+    pub fn from_cached_block(entry: CachableEntry<(HummockSstableId, u64), Box<Block>>) -> Self {
+        let ptr = entry.value().as_ref() as *const _;
+        Self {
+            _handle: BlockEntry::Cache(entry),
+            block: ptr,
+        }
+    }
+}
+
+impl Deref for BlockHolder {
+    type Target = Block;
+
+    fn deref(&self) -> &Self::Target {
+        unsafe { &(*self.block) }
+    }
+}
+
+unsafe impl Send for BlockHolder {}
+unsafe impl Sync for BlockHolder {}
+
+#[derive(PartialEq, Eq, Hash, Clone, Debug)]
+pub struct CacheKey {
+    sst: u64,
+    // TODO: To be compatible, use u64 for now. u32 is enough for idx.
+    idx: u64,
+}
+
+impl TieredCacheKey for CacheKey {
+    fn encoded_len() -> usize {
+        16
+    }
+
+    fn encode(&self, mut buf: &mut [u8]) {
+        buf.put_u64(self.sst);
+        buf.put_u64(self.idx);
+    }
+
+    fn decode(mut buf: &[u8]) -> Self {
+        let sst = buf.get_u64();
+        let idx = buf.get_u64();
+        Self { sst, idx }
+    }
+}
+
+// N (4B) + N * restart point len (N * 4B) + data len
+fn encode_block(block: &Block) -> Vec<u8> {
+    let mut buf = Vec::with_capacity(block.len() + 4 + block.restart_point_len() * 4);
+    buf.put_u32(block.restart_point_len() as u32);
+    for i in 0..block.restart_point_len() {
+        buf.put_u32(block.restart_point(i));
+    }
+    buf.extend_from_slice(block.data());
+    buf
+}
+
+fn decode_block(buf: &[u8]) -> Block {
+    let mut cursor = 0;
+
+    let restart_points_len = (&buf[cursor..cursor + 4]).get_u32() as usize;
+    cursor += 4;
+
+    let mut restart_points = Vec::with_capacity(restart_points_len);
+    for _ in 0..restart_points_len {
+        let restart_point = (&buf[cursor..cursor + 4]).get_u32();
+        cursor += 4;
+        restart_points.push(restart_point);
+    }
+
+    let data = Bytes::copy_from_slice(&buf[cursor..]);
+
+    Block::new(data, restart_points)
+}
+
+pub struct MemoryBlockCacheEventListener {
+    tiered_cache: TieredCache<CacheKey>,
+}
+
+impl LruCacheEventListener for MemoryBlockCacheEventListener {
+    type K = (HummockSstableId, u64);
+    type T = Box<Block>;
+
+    fn on_evict(&self, key: &Self::K, value: &Self::T) {
```

Comment:
The current LruCache holds a lock to do all the operation. The `encode_block` method seems heavy. I am not sure whether this will affect the performance of LRU cache.

Reply:
We may dump the evicted value to the buffer of tiered cache directly without any extra computation and only do the encoding when we write to file.

Reply:
> We may dump the evicted value to the buffer of tiered cache directly without any extra computation and only do the encoding when we write to file.

Agreed, I'll refactor that.